### PR TITLE
Add token_tree_to_xxx functions

### PR DESCRIPTION
**Pull request by edwin0cheng**
<del>As discus in  PR #1147 , this PR added a `mbe::MacroKind` .
Currently only 2 kind of macro are supported, `SourceFile` and `Block`.</del>

Added following functions for `tt::TokenTree` and `ast::Node` conversion:

* token_tree_to_expr
* token_tree_to_pat
* token_tree_to_ty
* token_tree_to_macro_stmts
* token_tree_to_macro_items

And added two new syntax kind:

* MACRO_ITEMS
* MACRO_STMTS

### CODE REVIEW

File: crates/ra_mbe/src/syntax_bridge.rs [hunk 1]
```diff
@@ -30,13 +59,26 @@ pub fn syntax_node_to_token_tree(node: &SyntaxNode) -> Option<(tt::Subtree, Toke
...
+) -> Result<MacroExpandResult, crate::ExpandError> {
     let token_source = SubtreeTokenSource::new(tt);
     let mut tree_sink = TtTreeSink::new(token_source.querier());
     ra_parser::parse(&token_source, &mut tree_sink);
     let syntax = tree_sink.inner.finish();
-    ast::SourceFile::cast(&syntax).unwrap().to_owned()
+
+    let res: Option<_> = match kind {
+        MacroKind::SourceFile => {
+            ast::SourceFile::cast(&syntax).map(|m| MacroExpandResult::SourceFile(m.to_owned()))
```
Comment (matklad) [inline crates/ra_mbe/src/syntax_bridge.rs]: The `ra_parser::parse` function should be inside this arm. `kind` should identify the parsing function we call, it should be `parse` for list of items, `expr` for expression, etc.

Let's also consult rustc to get the actual list of things macro can be parsed into: I don't think `MacroKind::Block` actually exist in the langauge?

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Here is the full list of rustc of non terminal iternoplation:
```
 NtItem(P<ast::Item>),
    NtBlock(P<ast::Block>),
    NtStmt(ast::Stmt),
    NtPat(P<ast::Pat>),
    NtExpr(P<ast::Expr>),
    NtTy(P<ast::Ty>),
    NtIdent(ast::Ident, /* is_raw */ bool),
    NtLifetime(ast::Ident),
    NtLiteral(P<ast::Expr>),
    /// Stuff inside brackets for attributes
    NtMeta(ast::MetaItem),
    NtPath(ast::Path),
    NtVis(ast::Visibility),
    NtTT(TokenTree),
    // These are not exposed to macros, but are used by quasiquote.
    NtArm(ast::Arm),
    NtImplItem(ast::ImplItem),
    NtTraitItem(ast::TraitItem),
    NtForeignItem(ast::ForeignItem),
    NtGenerics(ast::Generics),
    NtWhereClause(ast::WhereClause),
    NtArg(ast::Arg),
```
Give me some time to find where is the code to show what can be parsed into :)

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Should be : 
```
Expr(P<ast::Expr>) { "expression"; one fn visit_expr; fn visit_expr; fn make_expr; }
    Pat(P<ast::Pat>) { "pattern"; one fn visit_pat; fn visit_pat; fn make_pat; }
    Ty(P<ast::Ty>) { "type"; one fn visit_ty; fn visit_ty; fn make_ty; }
    Stmts(SmallVec<[ast::Stmt; 1]>) {
        "statement"; many fn flat_map_stmt; fn visit_stmt; fn make_stmts;
    }
    Items(SmallVec<[P<ast::Item>; 1]>) {
        "item"; many fn flat_map_item; fn visit_item; fn make_items;
    }
    TraitItems(SmallVec<[ast::TraitItem; 1]>) {
        "trait item"; many fn flat_map_trait_item; fn visit_trait_item; fn make_trait_items;
    }
    ImplItems(SmallVec<[ast::ImplItem; 1]>) {
        "impl item"; many fn flat_map_impl_item; fn visit_impl_item; fn make_impl_items;
    }
    ForeignItems(SmallVec<[ast::ForeignItem; 1]>) {
        "foreign item"; many fn flat_map_foreign_item; fn visit_foreign_item; fn make_foreign_items;
}
```

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: Heh, that's larger than I've hoped for :) I wonder how rustc handles the fact that a macro can expand to multiple items, but only to a single expression...

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: Aha, an error message in this case is illuminating: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=335e9a05f9629ac39c98cd61b5ac065b

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Haha, maybe you should file a bug in rustc :)
Anyway, as i remember correclty, we do not have a `ast::Stmt` type, we use `expr` , right?
```
TraitItems(SmallVec<[ast::TraitItem; 1]>) 
ImplItems(SmallVec<[ast::ImplItem; 1]>)
ForeignItems(SmallVec<[ast::ForeignItem; 1]>
```
Do we really need it ? Or we can use `Items` instead ?

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: We have `ExprStmt`, which is not `Expr`. 

>Do we really need these ? Or we can use Items instead ?

Given that there are quite a few variants there, I wonder if we can start with just `Expr`, `Items` and `Stmts`? I think it makes sense to start small, figure out the required API throughout the stack, and then fill the missing cases.

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Im not so familiar about `hir` side, What result type would you think these api should return ?
Something Like these ?
```rust
pub enum MacroExpandResult {
    Expr(TreeArc<ast::Expr>),
    Pat(TreeArc<ast::Pat>),
    Ty(TreeArc<ast::Ty>),
    Stmts(Vec<TreeArc<ast::ExprStmt>>),
    Items(Vec<TreeArc<ast::Item>>),
}
```
However we do not have `ast::Item` and `ast::Ty` type, we only have `ast::AstToken` and `ast::ItemList` type.

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: `ast::Item` is I think `ast::ModuleItem`, and `Ty` is a `TypeRef`.

>Something Like these ?

Hm, not exctly: it's the caller who decides what macro expands do, so it should be a bunch of functions like

```
tt_to_expr() -> TreeArc<ast::Expr>
tt_to_pat() -> TreeArc<ast::Pat>
```

For `Stmts` and `Items`, I *think* we want to return a single container node. IT probably shouldn't be `SourceFile`, we should introduce dedicated `MacroItems` and `MacroStmts`

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: So `MacroItems` and `MacroStmts` would be a new ast::Node and i need to add it in `grammer.ron`, right ?

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: right!

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: (the whole "parse tt into SourceFile" was just a hack, the proper soln are dedicated nodes)

**Comment by edwin0cheng** [comment]
Just remove the `MacroKind` enum , we do not need this as we separate each types on its own functions.
And i change all mbe internal tests to use `token_tree_to_macro_items`. 

The `token_tree_to_source_tree` should be removed in next PR after we figure out how to handle hir `SourceFile` ID related stuffs.

**Comment by edwin0cheng** [comment]
Foreget to return 'Result' instead of unwrap the `as::Node` to return, just fixed.

**Comment by matklad** [comment]
bors r+

Excellent!

**Pull request merged**